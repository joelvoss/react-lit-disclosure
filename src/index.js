import * as React from 'react';
import {
	useStableCallback,
	createNamedContext,
	makeId,
	useComposeRefs,
	composeEventHandlers,
} from '@react-lit/helper';
import { useId } from '@react-lit/auto-id';

////////////////////////////////////////////////////////////////////////////////

/**
 * @typedef {Object} DisclosureStates
 * @prop {string} OPEN
 * @prop {string} COLLAPSED
 */
const DisclosureStates = {
	OPEN: 'open',
	COLLAPSED: 'collapsed',
};

////////////////////////////////////////////////////////////////////////////////

/**
 * @typedef {Object} DisclosureContextValue
 * @prop {string} disclosureId
 * @prop {void} onSelect
 * @prop {boolean} open
 * @prop {string} panelId
 */

const DisclosureContext = createNamedContext('DisclosureContext', {});

////////////////////////////////////////////////////////////////////////////////

/**
 * A hook that exposes data for a given `Disclosure` component to it's
 * descendants.
 * @returns {{id: string, panelId: string, open: boolean }}
 */
export function useDisclosureContext() {
	/** @type {DisclosureContextValue} */
	let { open, panelId, disclosureId } = React.useContext(DisclosureContext);
	return React.useMemo(
		() => ({
			id: disclosureId,
			panelId,
			open,
		}),
		[disclosureId, open, panelId],
	);
}

////////////////////////////////////////////////////////////////////////////////

/**
 * Disclosure renders wrapper component and context provider for a
 * disclosure's button and
 * panel components. A disclosure should only have one button and one panel
 * descendant.
 */
export const Disclosure = ({
	children,
	defaultOpen = false,
	onChange,
	open: openProp,
	...props
}) => {
	const { current: isControlled } = React.useRef(openProp != null);

	const id =
		useId(props.id != null ? String(props.id) : undefined) || 'disclosure';
	const panelId = makeId('panel', id);

	const [open, setOpen] = React.useState(isControlled ? openProp : defaultOpen);
	const stableOnChange = useStableCallback(onChange);

	const onSelect = React.useCallback(() => {
		stableOnChange();
		if (!isControlled) {
			setOpen(open => !open);
		}
	}, [stableOnChange, isControlled]);

	const context = React.useMemo(
		() => ({
			disclosureId: id,
			onSelect,
			open,
			panelId,
		}),
		[onSelect, id, open, panelId],
	);

	if (isControlled && openProp !== open) {
		//NOTE(joel): If the component is controlled, we'll sync internal state
		// with the controlled state.
		setOpen(openProp);
	}

	return (
		<DisclosureContext.Provider value={context}>
			{children}
		</DisclosureContext.Provider>
	);
};

////////////////////////////////////////////////////////////////////////////////

/**
 * DisclosureButton renders the trigger button a user clicks to interact with
 * a disclosure.
 */
export const DisclosureButton = React.forwardRef(
	(
		{
			as: Comp = 'button',
			children,
			onClick,
			onMouseDown,
			onPointerDown,
			...props
		},
		parentRef,
	) => {
		/** @type {DisclosureContextValue} */
		const { onSelect, open, panelId } = React.useContext(DisclosureContext);
		const ownRef = React.useRef(null);

		const ref = useComposeRefs(parentRef, ownRef);

		/**
		 * handleClick
		 * @param {React.MouseEvent} event
		 */
		function handleClick(event) {
			event.preventDefault();
			ownRef.current && ownRef.current.focus();
			onSelect();
		}

		return (
			<Comp
				aria-controls={panelId}
				aria-expanded={open}
				{...props}
				data-state={open ? DisclosureStates.OPEN : DisclosureStates.COLLAPSED}
				ref={ref}
				onClick={composeEventHandlers(onClick, handleClick)}
			>
				{children}
			</Comp>
		);
	},
);

////////////////////////////////////////////////////////////////////////////////

/**
 * DisclosurePanel renders the collapsible panel in which inner content for an
 * disclosure item is rendered.
 */
export const DisclosurePanel = React.forwardRef(
	({ as: Comp = 'div', children, ...props }, parentRef) => {
		/** @type {DisclosureContextValue} */
		const { panelId, open } = React.useContext(DisclosureContext);

		return (
			<Comp
				ref={parentRef}
				hidden={!open}
				{...props}
				data-state={open ? DisclosureStates.OPEN : DisclosureStates.COLLAPSED}
				id={panelId}
				tabIndex={-1}
			>
				{children}
			</Comp>
		);
	},
);
