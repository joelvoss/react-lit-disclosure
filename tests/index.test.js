import * as React from 'react';
import { render, userEvent } from './test-utils';

import { Disclosure, DisclosureButton, DisclosurePanel } from '../src/index';

describe('<Disclosure />', () => {
	const Comp = props => {
		return (
			<Disclosure {...props}>
				<DisclosureButton>Button</DisclosureButton>
				<DisclosurePanel>Panel</DisclosurePanel>
			</Disclosure>
		);
	};

	it('should not have ARIA violations', async () => {
		let { container, getByRole } = render(<Comp />);

		await expect(container).toHaveNoAxeViolations();

		await userEvent.click(getByRole('button'));

		await expect(container).toHaveNoAxeViolations();
	});

	it('should render proper HTML', async () => {
		const { container } = render(<Comp />);
		expect(container.innerHTML).toMatchSnapshot();
	});

	it('should hide the panel content by default', async () => {
		const { getByText } = render(<Comp />);
		expect(getByText(/panel/i)).not.toBeVisible();
	});

	it(`should show the panel when defaultOpen is 'true'`, async () => {
		const { getByText } = render(<Comp defaultOpen />);
		expect(getByText(/panel/i)).toBeVisible();
	});

	it(`should hide the panel when defaultOpen is 'false'`, async () => {
		const { getByText } = render(<Comp defaultOpen={false} />);
		expect(getByText(/panel/i)).not.toBeVisible();
	});

	it(`should show the panel when open is 'true'`, async () => {
		const { getByText } = render(<Comp open />);
		expect(getByText(/panel/i)).toBeVisible();
	});

	it(`should hide the panel when open is 'false'`, async () => {
		const { getByText } = render(<Comp open={false} />);
		expect(getByText(/panel/i)).not.toBeVisible();
	});

	it(`should accept a custom id`, async () => {
		const { getByText } = render(<Comp id="custom-id" />);
		expect(getByText(/panel/i)).toHaveAttribute('id', 'panel--custom-id');
	});

	it(`should remove the panel from the navigation flow`, async () => {
		const { getByText } = render(<Comp />);
		expect(getByText(/panel/i)).toHaveAttribute('tabindex', '-1');
	});

	it('should set the correct aria attributes when collapsed', async () => {
		const { getByText } = render(<Comp />);
		let button = getByText(/button/i);
		let panel = getByText(/panel/i);

		expect(button).toHaveAttribute('aria-expanded', 'false');
		expect(button).toHaveAttribute('data-state', 'collapsed');

		expect(panel).toHaveAttribute('data-state', 'collapsed');
		expect(panel).toHaveAttribute('hidden');
	});

	it('should set the correct aria attributes when open', async () => {
		const { getByText } = render(<Comp open />);
		let button = getByText(/button/i);
		let panel = getByText(/panel/i);

		expect(button).toHaveAttribute('aria-expanded', 'true');
		expect(button).toHaveAttribute('data-state', 'open');

		expect(panel).toHaveAttribute('data-state', 'open');
		expect(panel).not.toHaveAttribute('hidden');
	});

	it(`should toggle on click`, async () => {
		const { getByText } = render(<Comp />);
		expect(getByText(/panel/i)).not.toBeVisible();

		await userEvent.click(getByText(/button/i));

		expect(getByText(/panel/i)).toBeVisible();
	});

	it(`should call 'onChange' on click`, async () => {
		let callback = jest.fn();
		const { getByText } = render(<Comp onChange={callback} />);

		await userEvent.click(getByText(/button/i));
		expect(callback).toHaveBeenCalledTimes(1);
		await userEvent.click(getByText(/button/i));
		expect(callback).toHaveBeenCalledTimes(2);
	});

	it(`should toggle on spacebar`, async () => {
		const { getByText } = render(<Comp />);

		getByText(/button/i).focus();

		await userEvent.keyboard(' ');
		expect(getByText(/panel/i)).toBeVisible();
		await userEvent.keyboard(' ');
		expect(getByText(/panel/i)).not.toBeVisible();
	});
});
