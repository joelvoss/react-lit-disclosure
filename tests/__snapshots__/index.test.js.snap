// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Disclosure /> should render proper HTML 1`] = `"<button aria-controls="panel--4" aria-expanded="false" data-state="collapsed">Button</button><div hidden="" data-state="collapsed" id="panel--4" tabindex="-1">Panel</div>"`;
